### ISSUE TITLE
`-Wunused` false positive on constexpr function used to compute return type

### ISSUE THREAD

**Issue by neildhar** [llvm/llvm-project#196564]
The following produces a warning ([godbolt](https://godbolt.org/z/crKW95456)):

```
  #include <type_traits>

  static constexpr bool returnInt(int) { return true; }

  template <typename T>
  std::conditional_t<returnInt(T{}), int, float> make() {
    return T{};
  }

  int makeInt() { return make<int>(); }
  float makeFloat() { return make<float>(); }
```

Warning:
```
function 'returnInt' is not needed and will not be emitted
```
